lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+  auto it = BBInfoMap.find(BB);
+  assert(it != BBInfoMap.end());
+  return it->second;
+}
+
+FieldReferenceGraph::Node *
+FieldReferenceGraph::createNewNode(FieldNumType FieldNum, DataBytesType S) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_FRG, dbgs() << "Create Node #" << NodeList.size()
+                                         << " for field " << FieldNum << "\n");
+  auto *Node = new FieldReferenceGraph::Node(NodeList.size(), FieldNum, S);
+  // Use NodeList to collect all allocated Nodes. Will be deallocated in
+  // destructor to avoid memory leak
+  NodeList.push_back(Node);
+  return Node;
+}
+
+// Connect two nodes by creating an Edge between them and set the weight to
+// be (C, D). The only tricky case is when the two nodes are already connected
+// by another edge, instead of creating a new edge, we update its weight.
+void FieldReferenceGraph::connectNodes(FieldReferenceGraph::Node *From,
+                                       FieldReferenceGraph::Node *To,
+                                       ExecutionCountType C, DataBytesType D,
+                                       bool BackEdge) {
+  if (From == To) {
+    // Don't connect a node back to itself to avoid a node never be able to
+    // collapse
+    return;
+  }
+  // Try to find existing edge between From and To
+  Edge *ExistEdge = NULL;
+  for (auto *E : From->OutEdges) {
+    if (E->ToNode == To)
+      ExistEdge = E;
+  }
+  if (ExistEdge) {
+    // If there's already an edge between the two nodes, only increase count and
+    // distance but not create a new edge
+    DEBUG_WITH_TYPE(
+        DEBUG_TYPE_FRG,
+        dbgs() << "Connect two nodes that already have a connection\n");
+    assert(BackEdge == ExistEdge->LoopArc);
+    updatePairByMerging(ExistEdge->ExecutionCount, ExistEdge->DataSize, C, D);
+    From->OutSum += C;
+    To->InSum += C;
+    return;
+  }
+  auto *Edge = new FieldReferenceGraph::Edge(EdgeList.size(), C, D);
+  EdgeList.push_back(Edge);
+  Edge->connectNodes(From, To);
+  From->OutEdges.insert(Edge);
+  To->InEdges.insert(Edge);
+  From->OutSum += C;
+  To->InSum += C;
+  if (BackEdge)
+    Edge->LoopArc = true;
+}
+
+// Collapse information of a Node to an Edge as a collapsed Entry
+void FieldReferenceGraph::collapseNodeToEdge(FieldReferenceGraph::Node *N,
+                                             FieldReferenceGraph::Edge *E) {
+  assert(E->Collapsed == false);
+  E->Collapsed = true;
+  if (N->FieldNum == 0)
+    return;
+  auto *Entry = new FieldReferenceGraph::Entry(EntryList.size(), N->FieldNum,
+                                               E->ExecutionCount, E->DataSize);
+  E->CollapsedEntries.insert(Entry);
+  EntryList.push_back(Entry);
+}
+
+// Copy collapsed Entry from one Edge to another
+void FieldReferenceGraph::moveCollapsedEntryToEdge(
+    FieldReferenceGraph::Entry *Entry, FieldReferenceGraph::Edge *FromEdge,
+    FieldReferenceGraph::Edge *ToEdge) {
+  assert(FromEdge->Collapsed);
```

Comment:
I passed in FromEdge because I intended to move the entry from one edge to another. If I copy, I think I probably don't need it. Will fix.